File: src/java/magma/app/compile/type/FunctionType.java
package magma.app.compile.type;

import magma.api.Tuple2;
import magma.api.collect.Joiner;
import magma.api.collect.list.Iterable;
import magma.app.TypeCompiler;

public record FunctionType(Iterable<String> args, String returns) implements Type {
    public String generateType() {
        var joinedArguments = this.args
                .iterWithIndices()
                .map((Tuple2<Integer, String> tuple) -> {
                    return "arg" + tuple.left() + " : " + tuple.right();
                })
                .collect(new Joiner(", "))
                .orElse("");

        return "(" + joinedArguments + ") => " + this.returns;
    }

    public boolean isFunctional() {
        return true;
    }

    public boolean isVar() {
        return false;
    }

    public String generateBeforeName() {
        return "";
    }

    public String generateSimple() {
        return TypeCompiler.generateType(this);
    }

    public boolean is(String type) {
        return "functional".equals(type);
    }
}

File: src/ts/magma/app/compile/type/FunctionType.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Iterable } from "../../../../magma/api/collect/list/Iterable";
import { Tuple2 } from "../../../../magma/api/Tuple2";
import { Joiner } from "../../../../magma/api/collect/Joiner";
import { TypeCompiler } from "../../../../magma/app/TypeCompiler";
export class FunctionType implements Node {
	args: Iterable<string>;
	returns: string;
	constructor (args: Iterable<string>, returns: string) {
		this.args = args;
		this.returns = returns;
	}
	generateNode(): string {
		let joinedArguments = this.args.iterWithIndices().map((tuple: Tuple2<number, string>) => {
			return "arg" + tuple.left() + " : " + tuple.right()/*unknown*/;
		}).collect(new Joiner(", ")).orElse("")/*unknown*/;
		return "(" + joinedArguments + ") => " + this.returns/*unknown*/;
	}
	isFunctional(): boolean {
		return true/*unknown*/;
	}
	isVar(): boolean {
		return false/*unknown*/;
	}
	generateBeforeName(): string {
		return ""/*unknown*/;
	}
	generateSimple(): string {
		return TypeCompiler.generateNode(this)/*unknown*/;
	}
	is(type: string): boolean {
		return "functional".equals(type)/*unknown*/;
	}
}
